src/yaml/to_yaml.cr

```diff
@@ -52,7 +52,19 @@ end
 
 class String
   def to_yaml(yaml : YAML::Builder)
-    yaml.scalar self
+    style = LibYAML::ScalarStyle::PLAIN
+    reserved = YAML::NULL_VALUES.includes?(self) ||
+               YAML::BOOL_VALUES.includes?(self) ||
+               YAML::INFINITY_VALUES.includes?(self.lchop('+')) ||
+               YAML::INFINITY_VALUES.includes?(self.lchop('-')) ||
+               YAML::NAN_VALUES.includes?(self) ||
+               self.to_i64?(underscore: true, prefix: true) ||
+               self.to_f64? ||
+               begin
+                 Time::Format::ISO_8601_DATE_TIME.parse(self)
+               rescue Time::Format::Error
+               end
```

Comment:
What about pushing the possible values for `-` and `+` infinity into `YAML::INFINITY_VALUES`? That would avoid chopping the string twice. Maybe introduce a `YAML::RESERVED` constant that would merge the INFINITY, NAN and BOOL values, which would help cleanup this method — as well as `String.new(YAML::PullParser)` above. We can also safely drop the `to_i?` call since `to_f?` will pass for both integers and floats.

Idealy we'd have a `Time.iso8601?` parser, too, but that will have to wait.

```crystal
if YAML::RESERVED.includes?(self) || to_f64? ||
    (Time::Format::ISO_8601_DATE_TIME.parse(self) rescue nil)
  style = LibYAML::ScalarStyle::DOUBLE_QUOTED
else
  style = LibYAML::ScalarStyle::ANY
end
yaml.scalar self, style
```

BTW: the `self.to_i64?` call should pass `underscore: false`.

Reply:
@ysbaddaden why drop underscore, and also, to_i also uses `prefix: true` which allows for the binary and hex numbers.